File: solutions/c/high-scores/1/high_scores.c
#include "high_scores.h"

int32_t latest(const int32_t* scores, size_t scores_len)
{
    return scores[scores_len - 1];
}

int32_t personal_best(const int32_t* scores, size_t scores_len)
{
    int32_t result = INT32_MIN;
    size_t i;
    for (i = 0; i < scores_len; ++i) {
        if (scores[i] > result) {
            result = scores[i];
        }
    }
    return result;
}

size_t personal_top_three(const int32_t* scores, size_t scores_len, int32_t* output)
{
    int32_t first = INT32_MIN;
    int32_t second = INT32_MIN;
    int32_t third = INT32_MIN;
    size_t i;
    size_t count = 0;
    for (i = 0; i < scores_len; ++i) {
        if (scores[i] > first) {
            third = second;
            second = first;
            first = scores[i];
        } else if (scores[i] > second) {
            third = second;
            second = scores[i];
        } else if (scores[i] > third) {
            third = scores[i];
        }
    }

    if (scores_len == 0) {
        return 0;
    }

    if (scores_len >= 1) {
        output[count++] = first;
    }
    if (scores_len >= 2) {
        output[count++] = second;
    }
    if (scores_len >= 3) {
        output[count++] = third;
    }

    return count;
}

File: solutions/java/high-scores/1/src/main/java/HighScores.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class HighScores {

    private final List<Integer> highScores;

    public HighScores(List<Integer> highScores) {
        this.highScores = highScores;
    }

    List<Integer> scores() {
        return List.copyOf(highScores);
    }

    Integer latest() {
        return highScores.getLast();
    }

    Integer personalBest() {
        var result = Integer.MIN_VALUE;
        for (var score : highScores) {
            if (score > result) {
                result = score;
            }
        }

        return result;
    }

    List<Integer> personalTopThree() {
        int[] sorted = new int[highScores.size()];
        for (int i = 0; i < highScores.size(); i++) {
            sorted[i] = highScores.get(i);
        }

        Arrays.sort(sorted);

        var result = new ArrayList<Integer>();
        int count = Math.min(3, sorted.length);
        for (int i = 0; i < count; i++) {
            result.add(sorted[sorted.length - 1 - i]);
        }

        return result;
    }

}
